compiler/rustc_hir_analysis/src/coherence/orphan.rs

```diff
@@ -462,81 +463,88 @@ fn emit_orphan_check_error<'tcx>(
+                    span,
+                    kind: ("an", "opaque type"),
+                    ty: opaque,
+                    local_ty,
+                }));
+            }
+            reported.unwrap()
         }
     }
 }
 
 fn lint_uncovered_ty_params<'tcx>(
     tcx: TyCtxt<'tcx>,
-    UncoveredTyParams { uncovered, local_ty }: UncoveredTyParams<TyCtxt<'tcx>, FxIndexSet<DefId>>,
+    data: UncoveredTy<TyCtxt<'tcx>, UncoveredTys<'tcx>>,
     impl_def_id: LocalDefId,
 ) {
     let hir_id = tcx.local_def_id_to_hir_id(impl_def_id);
 
-    for param_def_id in uncovered {
+    for param_def_id in data.uncovered.params {
         let span = tcx.def_ident_span(param_def_id).unwrap();
-        let name = tcx.item_name(param_def_id);
-
-        match local_ty {
-            Some(local_type) => tcx.emit_node_span_lint(
-                UNCOVERED_PARAM_IN_PROJECTION,
-                hir_id,
-                span,
-                errors::TyParamFirstLocalLint { span, note: (), param: name, local_type },
-            ),
-            None => tcx.emit_node_span_lint(
-                UNCOVERED_PARAM_IN_PROJECTION,
-                hir_id,
-                span,
-                errors::TyParamSomeLint { span, note: (), param: name },
-            ),
-        };
+        tcx.emit_node_span_lint(UNCOVERED_PARAM_IN_PROJECTION, hir_id, span, errors::UncoveredTy {
+            span,
+            kind: ("a", "type parameter"),
+            ty: tcx.item_name(param_def_id),
+            local_ty: data.local_ty,
+        });
     }
 }
 
-struct UncoveredTyParamCollector<'cx, 'tcx> {
+struct UncoveredTyCollector<'cx, 'tcx> {
     infcx: &'cx InferCtxt<'tcx>,
-    uncovered_params: FxIndexSet<DefId>,
+    uncovered: UncoveredTys<'tcx>,
 }
 
-impl<'tcx> TypeVisitor<TyCtxt<'tcx>> for UncoveredTyParamCollector<'_, 'tcx> {
+impl<'tcx> TypeVisitor<TyCtxt<'tcx>> for UncoveredTyCollector<'_, 'tcx> {
     fn visit_ty(&mut self, ty: Ty<'tcx>) -> Self::Result {
-        if !ty.has_type_flags(ty::TypeFlags::HAS_TY_INFER) {
+        if !ty.has_type_flags(TypeFlags::HAS_TY_INFER | TypeFlags::HAS_ALIAS) {
             return;
         }
-        let ty::Infer(ty::TyVar(vid)) = *ty.kind() else {
-            return ty.super_visit_with(self);
-        };
-        let origin = self.infcx.type_var_origin(vid);
-        if let Some(def_id) = origin.param_def_id {
-            self.uncovered_params.insert(def_id);
+        match *ty.kind() {
+            ty::Infer(ty::TyVar(vid)) => {
+                if let Some(def_id) = self.infcx.type_var_origin(vid).param_def_id {
+                    self.uncovered.params.insert(def_id);
+                }
+            }
+            // FIXME(fmease): This is a temporary HACK; rework it. In the next solver,
+            // we normalize structurally meaning opaque types "remain behind" weak
+            // aliases. I kind of want to expand_weak_alias_tys this but that's
+            // pretty hacky, too.
```

Comment:
The HACK mentioned in the PR description.

Reply:
Yeah :thinking: I don't really know what to do here. I feel like ideally we potentially don't need to recompute things here and report an error using only the original uncovered type as this folder is somewhat :/

I am personally fine with slightly regressing diagnostics here if necessary for that, but don't have a clear idea of how bad this would be